opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -0,0 +1,343 @@
+
+    @abstractmethod
+    def create_observable_up_down_counter(
+        self, name, callback, unit="", description=""
+    ) -> ObservableUpDownCounter:
+        with self._lock:
+            self._check_instrument_name(name)
+
+
+class ProxyMeter(Meter):
+    def __init__(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ):
+        super().__init__(name, version=version, schema_url=schema_url)
+        self._lock = RLock()
+        self._real_meter: Optional[Meter] = None
+        self._noop_meter = _DefaultMeter(
+            name, version=version, schema_url=schema_url
+        )
+
+    @property
+    def _meter(self) -> Meter:
+        with self._lock:
+            if self._real_meter is not None:
+                return self._real_meter
+
+            if _METER_PROVIDER:
+                self._real_meter = _METER_PROVIDER.get_meter(
+                    self._name,
+                    self._version,
+                )
+                return self._real_meter
+            return self._noop_meter
+
+    def create_counter(self, *args, **kwargs) -> Counter:
+        with self._lock:
+            super().create_counter(*args, **kwargs)
+            return self._meter.create_counter(*args, **kwargs)
+
+    def create_up_down_counter(self, *args, **kwargs) -> UpDownCounter:
+        with self._lock:
+            super().create_up_down_counter(*args, **kwargs)
+            return self._meter.create_up_down_counter(*args, **kwargs)
+
+    def create_observable_counter(self, *args, **kwargs) -> ObservableCounter:
+        with self._lock:
+            super().create_observable_counter(*args, **kwargs)
+            return self._meter.create_observable_counter(*args, **kwargs)
+
+    def create_histogram(self, *args, **kwargs) -> Histogram:
+        with self._lock:
+            super().create_histogram(*args, **kwargs)
+            return self._meter.create_histogram(*args, **kwargs)
+
+    def create_observable_gauge(self, *args, **kwargs) -> ObservableGauge:
+        with self._lock:
+            super().create_observable_gauge(*args, **kwargs)
+            return self._meter.create_observable_gauge(*args, **kwargs)
+
+    def create_observable_up_down_counter(
+        self, *args, **kwargs
+    ) -> ObservableUpDownCounter:
+        with self._lock:
+            super().create_observable_up_down_counter(*args, **kwargs)
+            return self._meter.create_observable_up_down_counter(
+                *args, **kwargs
+            )
+
+
+class _DefaultMeter(Meter):
+    def create_counter(self, name, unit="", description="") -> Counter:
+        with self._lock:
+            super().create_counter(name, unit=unit, description=description)
+            return DefaultCounter(name, unit=unit, description=description)
+
+    def create_up_down_counter(
+        self, name, unit="", description=""
+    ) -> UpDownCounter:
+        with self._lock:
+            super().create_up_down_counter(
```

Comment:
I think we should avoid this need-to-call-super pattern. I'd rather have a superclass method called `_validate()` because you can tell what it's doing. Another option is a template method so the "recipe" of what to call is in the super class but we can override bits of it. https://martinfowler.com/bliki/CallSuper.html

Reply:
No, having a `_validate` method means having an implementation detail of the parent class that the subclasses should not use.

We could use the suggested pattern, but that would mean less control for the subclass on when the validation happens. For this particular scenario it does not matter much, but it could be problematic in other situations. I also find it annoying having to call `super` but it is a tradeoff between convenience and control. We also use `super` elsewhere so it would be kind of inconsistent.

Reply:
Your superclass `create_up_down_counter()` isn't even returning an `UpDownCounter`. If it did, what would you do with it, just throw it away? The easiest solution is to just have a validation util that child classes can call.

> No, having a `_validate` method means having an implementation detail of the parent class that the subclasses should not use.

I don't see how this is the case. This is why languages like Java have `protected`, so subclasses can call parent class methods

Reply:
> Your superclass `create_up_down_counter()` isn't even returning an `UpDownCounter`. If it did, what would you do with it, just throw it away?

Hm, of course it is not returning any kind of `UpDownCounter` :thinking: it is an abstract method. Why should it do so? Imagine it was returning an `UpDownCounter`, if it did, then that object would be an instance of a concrete class, so which concrete class would be? Sorry, I don't really understand how this is an argument against this implementation :shrug: 

> The easiest solution is to just have a validation util that child classes can call.

We are just allowing classes to call `super().x()` in their own `x` methods that they must implement. This is no harder nor easier compared to having a different validation method but that is not the point. We are just using here the normal approach of having child classes call `super`.

> I don't see how this is the case. This is why languages like Java have `protected`, so subclasses can call parent class methods

Yes, sure, but Python does not have that feature, even when I would love to have it. :shrug:

Reply:
We have agreed on following the convention where a single leading underscore is considered `protected`.